README.md

     ______   _____    _______                    _  _       
    |  ___ \ (____ \  (_______)             _    (_)| |       
    | | _ | | _   \ \  ______   ___  _   _ | |_   _ | |  ___  
    | || || || |   | |(_____ \ (___)| | | ||  _) | || | /___)
    | || || || |__/ /  _____) )     | |_| || |__ | || ||___ |
    |_||_||_||_____/  (______/       \____| \___)|_||_|(___/  
               Copyright (C) 2009-2013 by BoboTiG

MD5-utils comes with ABSOLUTELY NO WARRANTY.  
This is free software, and you are welcome to redistribute it under  
certain conditions. See the GNU General Public Licence for details.  


English
---

**What is this?**

MD5-utils is a script for MD5 hash function manipulations.   
It can crypt one or more word(s) and try to crack one or more hash(es)  
using online servers. 

**Options**

You can change few options on the top of script.
	
**Examples**

* View help message:

	    perl md5-utils.pl -h
	
* Print server list:

	    perl md5-utils.pl -l
	
* Look for an update:

	    perl md5-utils.pl -u
	
* Interactive mode:

	    perl md5-utils.pl -i
	
* Hash one or more word(s):

	    perl md5-utils.pl -c <word> <word> ...
	
* Try to crack one or more hash(es):

	    perl md5-utils.pl -d <hash> <hash> ...
	
* Try to crack hashes from file:

	    perl md5-utils.pl -f way/to/the/file
	    (found-txt will contain reversible hashes, not-found.txt will contain others)

**Special uses**

You can choose to receive or send args by two ways:

* One for the crypt option, it send results to <STDOUT> and can be used by another  
script using <STDIN>, try this:

	    perl md5-utils.pl -c <word> <word> --stdout
	
* The second is for the crak option, it receive hashes from another script
using <STDIN>, try this:

	    perl md5-utils.pl -d --stdin


Français
---

**De quoi s'agit-il ?**

MD5-utils est un script pour la manipulation de hashes chiffrés en MD5.  
Il peut crypter un ou plusieurs mot(s) et tenter de cracker un ou plusieurs hash(es)  
en effectuant des recherches sur plusieurs serveurs en ligne.

**Options**

	Vous pouvez changer quelques options en début de script.

**Exemples**

* Afficher le message d'aide :
 
	    perl md5-utils.pl -h
	
* Afficher la liste des serveurs :

	    perl md5-utils.pl -l
	
* Voir si une mise à jour existe :

	    perl md5-utils.pl -u
	
* Mode intéractif :

	    perl md5-utils.pl -i
	
* Hasher un ou plusieurs mot(s) :

	    perl md5-utils.pl -c <mot> <mot> ...
	
* Tenter de cracker un ou plusieurs hash(es) :

	    perl md5-utils.pl -d <hash> <hash> ...
	
* Tenter de cracker les hash contenus dans un fichier :

	    perl md5-utils.pl -f chemin/vers/le/fichier
	    (found-txt contiendra les hashes réversibles, not-found.txt contiendra les autres)

**Utilisations spéciales**

Vous avez le choix entre deux autres moyens de recevoir et envoyer des informations :

* Une pour l'option de cryptage, à enverra tout à <STDOUT> pour être utilisé par un autre  
script qui bouclera sur <STDIN>, essayez ceci :

		perl md5-utils.pl -c <mot> <mot> --stdout
		
* La deuxième pour l'option de crack, elle recevra les hashes envoyés par un autre  
script en utilisant <STDIN>, essayez ceci :

		perl md5-utils.pl -d --stdin


End / Fin
---

use bye::bye qw(Tcho);


doc/CHANGELOG
=== 2013/xx/yy ===
+ server added: tobtu.com


=== 2013/06/28 ===
+ fix update routine
+ servers regexp fixed:
	www.md5this.com, www.stringfunction.com, www.google.com, md5.my-addr.com
	md5crack.com, www.bigtrapeze.com, www.md5hood.com, md5pass.info
+ servers commented: md5.rednoize.com
+ server deleted: 
	md5.bobotig.fr, tools.benramsey.com, www.rmd5.com, schwett.com
	alimamed.pp.ru, hashcracking.info, opencrack.hashkiller.com
	www.passcracking.com, milw0rm.com, www.shell-storm.org, www.hashchecker.com
	md5.allfact.info, bokehman.com, www.md5decrypter.com(captcha)


=== 2011/01/22 ===
+ added the MD5-utils lite version (md5crack)
+ code reviewed
+ server added: www.onlinehashcrack.com
+ server deleted: cryptohaze.com, www.gdataonline.com, blacklight.gotdns.org, decrypt.vanvan.cc


=== 2010/09/21 ===
+ server added: api.dev.c0llision.net
+ servers uncommented: hashcracking.info, www.passcracking.com, milw0rm.com, decrypt.vanvan.cc


=== 2010/07/28 ===
+ avoid warning message when a thread terminated abnormally
+ code modified to succesfully pass Perl::Critic (severity 'brutal')


=== 2010/06/25 ===
+ change error messages filtering by adding server name and its message into a hash %dont_mind
+ sanitizing configurations variables
+ change 0 and 1 to false and true into md5-utils.cfg
+ useless use of HTTP::Request::Common module
+ servers commented: milw0rm.com

=== 2010/06/19 === *** Upgrade to the GNU GPL license version 3 or higher ***
+ [TODO] In interactive mode, add history for GNU/Linux users: done.
+ [TODO] "MD5-utils is too polite" => need to launch several requests simultaneously: done.
+ md5-utils.cfg created to manage options
+ MD5utils::i18n and MD5utils::servers packages created for easier modifications
+ added proxy support
+ added coloration for board
+ added --imode option to enter into interactive mode
+ interactive mode reviewed, more complete and as similar as normal mode
+ added options to interactive mode:
    [] 'crypt' to crypt words like -c
    [] 'file' to crack hashes contained into a file like -f option
    [] 'set' to manage options
    [] 'quit' to exit
+ code reviewed:
	[] HTML::Entities, Switch, charnames and Encode are no longer needed ;
	[] check for all necessary modules ;
	[] fixed color codes ;
	[] with -f (or --file) option, files found.txt and not-found.txt will append into the same folder as the file containing MD hashes ;
	[] deleted searches for salted MD5 and file MD5 sums ;
	[] deleted -t (or --test) option ;
	[] added complete informations for all subs ;
	[] added username in interactive mode prompt ;
	[] added read_config() sub to read the configuration file ;
	[] added is_numeric() sub to test if a variable is numeric or not ;
	[] added purge_array() sub to purge an array from a value ;
+ servers added: www.google.com, www.stringfunction.com, tools.benramsey.com, schwett.com, md5.my-addr.com
+ servers regexp fixed: www.cloudcracker.net, hashcracking.info, www.md5decryption.com
+ servers deleted:
	[down]: ice.breaker.free.fr, www.md5cracker.pl, md5.hamaney.net, www.securitystats.com, md5.ip-domain.com.cn
	[captcha]: www.md5decrypter.co.uk
	[too bad performances]: www.cmd5.com, tmto.org


=== 2009/09/24 ===
+ added -i option to enter in interactive mode
    [] available commands are clear, crack, exit, help, list and update
+ added clear_console() sub to clear the console and add title
+ added message_interactive() sub to display a nice title in interactive mode (font used: stop)
+ added eradicate_accents() sub to eradicate accents if french language is chosen under Windows OS like
+ added MD5-utils.bat and MD5-utils short cuts
+ code reviewed, change loops 'for' to 'foreach' and other details
+ special server added for file md5 sum (sunsolve.sun.com)
+ server deleted cause of captcha (hash.insidepro.com)


=== 2009/08/20 === *** distributed with Joomscan ***
+ bug fixed when $method was out of range
+ bug fixed when $result contained russian chars
+ added arg for -s (or --salt), use '-1' for the method to test all available methods
+ delete wrong -s (or --salt) explanation in help message
+ when $result is a hash, we try to crack it too (not for -di option)
+ regexp fixed for www.passcracking.com and milw0rm.com servers
+ servers added (md5.allfact.info, bokehman.com, www.cloudcracker.net)


=== 2009/07/06 ===
+ bug fixed whith -f (or --file) option (solve end char problems with Mac and Windows)
+ bug fixed with -d option (length($hash_to_crack))
+ added rtrim sub (remove trailing withespace)


=== 2009/07/01 ===
+ added output file for -d option (founds.txt and nofounds.txt are permanent)
+ added -di option, it will return the plaintext found or nothing
+ added a variable on the top script for disabled servers, -l or --list to know which
+ is_in_array() sub reviewed, more efficient now
+ change check_response() sub to is_in_array() sub
+ servers added (md5.bobotig.fr, cryptohaze.com)


=== 2009/06/29 ===
+ when -f option is used, the file was delete, not now.
+ uniq_array() sub reviewed for returned just hashes not found
+ search() sub reviewed, now it add all not found hashes in the not-found.txt file
+ added -s and --salt options for trying to crack one hash with salt
+ change syntax in found.txt (hash:text)
+ added array containing sites which use a captcha, just for the furtur when will be able to bypass them


=== 2009/06/24 ===
+ when -f option is used, we sort and uniq the list
+ when -f option is used, if finished, we delete the file
+ uniq_array() sub reviewed, more efficient now
+ added --salt option, you can specify a salt (--salt="SALT")
+ added --method option to specify the method to use (--method=x)
+ added --show-methods option to know available methods
+ added some CMS and boards which used these methods (infos from Waraxe Forums)
+ server added (md5.hamaney.net) [SALT]


=== 2009/06/23 ===
+ bug fixed in console when $result contained accentued char


=== 2009/06/19 ===
+ bug fixed when $result contained HTML codes

test-servers.pl
#!/usr/bin/perl
#
# Test all servers.
#


### [ Modules ] ###
use 5.010;
use strict;
use Cwd qw(abs_path);
use Digest::MD5;
use Parallel::ForkManager;
use LWP::UserAgent;
use warnings;

# Personal modules, we need to locate them
my $location;
# Why -15 in the next line? -> lenght('test-servers.pl') = 15
BEGIN { $location = substr(abs_path($0), 0, -15); }
use lib $location.'/bin';
use MD5utils::servers;
### [ /Modules ] ###


### [ Script configuration ] ###
our $VERSION    = '2013.06.28';
my $max_threads = 10;
my $user_agent  = 'Mozilla/5.0 (X11; Linux x86_64; rv:8.0) Gecko/20100101 Firefox/8.0 Iceweasel/8.0';
my @forms;
### [ /Script configuration ] ###

#
# Fonction	: args
# Objectif	: arguments traitment
# Entries	: (array)arguments
# Returns	: void
# Update	: 20130628
#
sub args {
	# Retrieve all servers
	foreach my $server ( @servers::form_get ) {
		push @forms, ['get', $server];
	}
	foreach my $server ( @servers::form_post ) {
		push @forms, ['post', $server];
	}
	
	my $n = $#forms;
	my $pm = Parallel::ForkManager->new($max_threads);
	say 'Total servers : '.($n + 1);
	for my $i ( 0 .. $n ) {
		$pm->start($i) and next;
		search('202cb962ac59075b964b07152d234b70', $i); # 123
		$pm->finish();
	}
	# Servers with GET method
	$pm->wait_all_children;
	return 0;
}

#
# Fonction	: request
# Objectif	: launch the server resquest for one hash
# Entries	:
#		- (string)form method
#		- (string)hash
#		- (string)server_name
#		- (string)server_link
#		- (string)server_regexp
#		- (array)post arguments
# Returns	: void
# Update	: 20111205
#
sub request {
	my ($method, $hash, $server, $link, $regexp, %arguments) = @_;
	my $ua;
	my $requete;
	
	$ua = LWP::UserAgent->new();
	$ua->agent($user_agent);
	if ( $method eq 'get' ) {
		eval { $requete = $ua->get($link) };
	} else {
		eval { $requete = $ua->post($link, \%arguments) };
	}
	if ( $@ =~ /Can't connect to/ ) {
		printf ' %s : connection imp.'."\n", $server;
		return;
	}
	if ( $requete->is_success ) {
		$requete->content =~ m/($regexp)/ms;
		my ($result) = $+ || undef;
		if ( defined $result ) {
			search_result($hash, $result, $server);
		} else {
			printf ' %s : result undefined.'."\n", $server;
		}
	} else {
		printf ' %s : connection bad.'."\n", $server;
	}
	return;
}


#
# Fonction	: multi-threaded search
# Objectif	: try to crack one hash using one server
# Entries	:
#		- (string)hash
#		- (int)key
# Returns	: void
# Update	: 20111205
#
sub search {
	my ($hash, $i) = @_;
	my $method = $forms[$i][0];
	my $server = $forms[$i][1][0];
	my $link   = $forms[$i][1][1];
	my $regexp = $forms[$i][1][2];
	my %arguments;
	
	if ( $method eq 'get' ) {
		$link =~ s/\{hash\}/$hash/ms;
	} else {
		for my $j ( 1..$forms[$i][1]['3'] ) {
			my $arg0 = $forms[$i][1][(int '3' + $j)][0];
			my $arg1 = $forms[$i][1][(int '3' + $j)][1];
			if ( $arg1 =~ m/\{hash\}/ms ) {
				$arg1 =~ s/\{hash\}/$hash/ms;
			}
			$arguments{$arg0} = $arg1;
		}
	}
	request($method, $hash, $server, $link, $regexp, %arguments);
	return;
}


#
# Fonction	: search_result
# Objectif	: manage results
# Entries	:
#		- (string)hash
#		- (mixed)result
#		- (string)server
# Returns	: void
# Update	: 20111205
#
sub search_result {
	my ($hash, $result, $server) = @_;
	if ( $result ) {
		my $res = Digest::MD5::md5_hex($result);
		printf '%s=%s:%s', $hash, $res, $result;
		if ( $hash eq $res ) {
			printf " %s : %s\n", $server, $result;
		} else {
			printf " %s : bad result\n", $server;
		}
	} else {
		printf " %s : not found\n", $server;
	}
	return;
}
### [ /Subs ] ###


###
# Let's go!
exit args(@ARGV);


md5crack/md5-utils-lite.pl
#!/usr/bin/perl
#
# $Id       : MD5-utils-lite $
# $Author   : BoboTiG $
# $Revision : 2 $
# $Date     : 2011/12/05 $
# $HeadURL  : http://www.bobotig.fr/ $
#
# MD5-utils Lite version, returns :
#   - nothing if wrong hash or no result ;
#   - the MD5 reverse of the hash if found.
#
# Usage : perl md5-utils-lite.pl <hash>
#
# Dependances :
#    - Digest::MD5
#    - LWP::UserAgent
#


### [ Modules ] ###
use strict;
use Digest::MD5;
use LWP::UserAgent;
use warnings;
### [ /Modules ] ###


### [ Script configuration ] ###
our $VERSION    = '2011.12.05';
my $user_agent  = 'Mozilla/5.0 (X11; Linux x86_64; rv:8.0) Gecko/20100101 Firefox/8.0 Iceweasel/8.0';
my @forms = (
	['get', ['md5.bobotig.fr', 'http://md5.bobotig.fr/index.php?hash={hash}', '&rsaquo;&rsaquo;&rsaquo;<xmp class="result">(.+)</xmp>&lsaquo;&lsaquo;&lsaquo;']],
	['get', ['api.dev.c0llision.net', 'http://api.dev.c0llision.net/crack/md5/{hash}', '<raw>(.+)</raw>']],
	['get', ['md5.gromweb.com', 'http://md5.gromweb.com/?md5={hash}', '<td><input type="text" name="string" value="(.+)" id="form_string" maxlength="255" size="40" /></td>']],
	['get', ['md5.darkbyte.ru', 'http://md5.darkbyte.ru/api.php?q={hash}', '(.+)']],
	['get', ['www.stringfunction.com', 'http://www.stringfunction.com/md5-decrypter.html?s={hash}', '<textarea class="textarea-input-tool-b" rows="10" cols="50" name="result" id="textarea_md5_decrypter">(.+)</textarea>\n\t{10}']],
	['get', ['md5.hashcracking.com', 'http://md5.hashcracking.com/search.php?md5={hash}', 'Cleartext of [A-Fa-f0-9]{32} is (.+)']],
	['get', ['md5.rednoize.com', 'http://md5.rednoize.com/?q={hash}', '<div id="result" >(.+)</div>\n\t{3}']],
	['post', ['www.onlinehashcrack.com', 'http://www.onlinehashcrack.com/free-hash-reverse.php', '\sPlain text : <b style="letter-spacing:1.2px">(.+)</b><br />\s', 2, ['hashToSearch', '{hash}'], ['searchHash', 'Search']]],
	['post', ['www.md5decryption.com', 'http://www.md5decryption.com/index.php', '<font size=\'2\'>Decrypted Text: </b>(.+)</font><br/><center>', 2, ['hash', '{hash}'], ['submit', 'Decrypt It!']]],
	['post', ['www.cloudcracker.net', 'http://www.cloudcracker.net/index.php', 'Word: <input type="text" class="word" name="theword" readonly="" onclick="(.*)" value="(.+)" />', 1, ['inputbox', '{hash}']]],
	['post', ['www.md5hood.com', 'http://md5hood.com/index.php/cracker/crack', '<div class="result_true">(.+)</div>\s*</fieldset>', 2, ['md5', '{hash}'], ['submit', 'Go']]],
);
### [ /Script configuration ] ###


#-----------------------------------------------------------------------
# Fonction	: request
# Objectif	: launch the server resquest for one hash
# Entries	:
#		- (string)form method
#		- (string)hash
#		- (string)server_name
#		- (string)server_link
#		- (string)server_regexp
#		- (array)POST arguments
# Returns	: void
# Update	: 20111205
#
sub request {
	my ($method, $hash, $server, $link, $regexp, %arguments) = @_;
	my $ua;
	my $requete;
	
	$ua = LWP::UserAgent->new();
	$ua->agent($user_agent);
	if ( $method eq 'get' ) {
		eval { $requete = $ua->get($link) };
	} else {
		eval { $requete = $ua->post($link, \%arguments) };
	}
	if ( $@ =~ /Can't connect to/ || $requete->is_error ) { return; }
	$requete->content =~ m/($regexp)/ms;
	my ($result) = $+ || undef;
	if ( defined $result ) {
		if ( $hash eq Digest::MD5::md5_hex($result) ) {
			print $result;
			exit 0;
		}
	}
	return;
} # end request --------------------------------------------------------


#-----------------------------------------------------------------------
# Fonction	: search
# Objectif	: try to crack one hash using one server
# Entries	:
#		- (string)hash
#		- (int)key
# Returns	: void
# Update	: 20111205
#
sub search {
	my ($hash, $i) = @_;
	my $method     = $forms[$i][0];
	my $server     = $forms[$i][1][0];
	my $link       = $forms[$i][1][1];
	my $regexp     = $forms[$i][1][2];
	my %arguments;
	
	if ( $method eq 'get' ) {
		$link =~ s/\{hash\}/$hash/ms;
	} else {
		for my $j ( 1..$forms[$i][1]['3'] ) {
			my $arg0 = $forms[$i][1][(int '3' + $j)][0];
			my $arg1 = $forms[$i][1][(int '3' + $j)][1];
			if ( $arg1 =~ m/\{hash\}/ms ) {
				$arg1 =~ s/\{hash\}/$hash/ms;
			}
			$arguments{$arg0} = $arg1;
		}
	}
	return request($method, $hash, $server, $link, $regexp, %arguments);
} # end search ---------------------------------------------------------


###
# C'est parti mon kiki !
my $hash = lc shift || undef;
if ( $hash =~ m/^[0-9a-f]{32}$/ms ) {
	my $n = $#forms;
	for my $i ( 0 .. $n ) {
		search($hash, $i);
	}
}
exit 0;
